Add support version minecraft for 1.5.2-1.7.2

Gamer-VII opened the pull request.

Gamer-VII committed:
Simplify and extend file retrieval methods in ForgeVersion.

Refactored `GetInstallerFile` to use an expression-bodied member for clarity and conciseness. Added a new method `GetUniversalFile` to retrieve universal file types, improving functionality and code extensibility.

Gamer-VII committed:
Refactor Forge installer extraction logic.

Centralized duplicate extraction logic into a reusable method to improve code maintainability. Introduced support for universal installer handling and added `GetVanillaExtractor` for vanilla JAR extraction.

Gamer-VII committed:
Implement Forge installer for oldest Forge versions

Added logic to handle installation, libraries, and version JSON for the oldest Forge versions. Introduced utility methods for extracting and processing installer profiles, libraries, and JAR compression to ensure compatibility with legacy setups.

Gamer-VII committed:
Update installer versions and refine process initialization

Updated Minecraft and Forge versions to older configurations for compatibility testing. Replaced `CreateProcessAsync` with `BuildProcessAsync` to align with API changes. Added JVM argument to ignore invalid Minecraft certificates during runtime.

Gamer-VII commented:
https://github.com/CmlLib/CmlLib.Core/pull/121#issue-2970173272

These changes will only work after the Pull Request is accepted and the CmlLib package is updated to version 4.0.5.

AlphaBs commented on CmlLib.Core.Installer.Forge/Installers/FOldest.cs:
```diff
@@ -15,8 +18,108 @@ public FOldest(string versionName, ForgeVersion forgeVersion)
+        }
+    }
+
+    // private async Task extractUniversal(MinecraftPath path)
```
remove unused code

GamerVII-NET replied:
Ok

AlphaBs commented on SampleForgeInstaller/Program.cs:
```diff
@@ -41,9 +53,13 @@
     Session = MSession.CreateOfflineSession("TaiogStudio"),
+    ExtraJvmArguments = new []
+    {
+        new MArgument("-Dfml.ignoreInvalidMinecraftCertificates=true"),
```
can you explain what it is?

GamerVII-NET replied:
```
D:\Programs\JetBrains\Rider\plugins\dpa\DotFiles\JetBrains.DPA.Runner.exe --handle=18732 --backend-pid=9532 --etw-collect-flags=58720014 --detach-event-name=dpa.detach.9532.18 --refresh-interval=1 -- D:/Projects/RiderProjects/CmlLib.Core.Installer.Forge/SampleForgeInstaller/bin/Debug/net6.0/SampleForgeInstaller.exe
Apr 03, 2025 12:10:02 PM net.minecraft.launchwrapper.LogWrapper log
INFO: Loading tweak class name cpw.mods.fml.common.launcher.FMLTweaker
Apr 03, 2025 12:10:02 PM net.minecraft.launchwrapper.LogWrapper log
INFO: Using primary tweak class name cpw.mods.fml.common.launcher.FMLTweaker
Apr 03, 2025 12:10:02 PM net.minecraft.launchwrapper.LogWrapper log
INFO: Calling tweak class cpw.mods.fml.common.launcher.FMLTweaker
2025-04-03 12:10:02 [INFO] [ForgeModLoader] Forge Mod Loader version 6.4.50.1,345 for Minecraft 1.6.4 loading
2025-04-03 12:10:02 [INFO] [ForgeModLoader] Java is Java HotSpot(TM) 64-Bit Server VM, version 1.8.0_51, running on Windows 10:amd64:10.0, installed at C:\Users\a.a.terentiev\AppData\Roaming\.minecraft\runtime\windows-x64\jre-legacy
2025-04-03 12:10:02 [INFO] [ForgeModLoader] Loading tweak class name cpw.mods.fml.common.launcher.FMLInjectionAndSortingTweaker
2025-04-03 12:10:02 [INFO] [ForgeModLoader] Loading tweak class name cpw.mods.fml.common.launcher.FMLDeobfTweaker
2025-04-03 12:10:02 [INFO] [ForgeModLoader] Calling tweak class cpw.mods.fml.common.launcher.FMLInjectionAndSortingTweaker
2025-04-03 12:10:02 [INFO] [ForgeModLoader] Calling tweak class cpw.mods.fml.common.launcher.FMLInjectionAndSortingTweaker
2025-04-03 12:10:02 [INFO] [ForgeModLoader] Calling tweak class cpw.mods.fml.relauncher.CoreModManager$FMLPluginWrapper
2025-04-03 12:10:02 [INFO] [STDOUT] Loaded 40 rules from AccessTransformer config file fml_at.cfg
2025-04-03 12:10:03 [INFO] [ForgeModLoader] Found valid fingerprint for Minecraft Forge. Certificate fingerprint e3c3d50c7c986df74c645c0ac54639741c90a557
2025-04-03 12:10:03 [SEVERE] [ForgeModLoader] The minecraft jar file:/C:/Users/a.a.terentiev/AppData/Roaming/.minecraft/versions/1.6.4-Forge9.11.1.1345/1.6.4-Forge9.11.1.1345.jar!/net/minecraft/client/ClientBrandRetriever.class appears to be corrupt! There has been CRITICAL TAMPERING WITH MINECRAFT, it is highly unlikely minecraft will work! STOP NOW, get a clean copy and try again!
2025-04-03 12:10:03 [SEVERE] [ForgeModLoader] For your safety, FML will not launch minecraft. You will need to fetch a clean version of the minecraft jar file
2025-04-03 12:10:03 [SEVERE] [ForgeModLoader] Technical information: The class net.minecraft.client.ClientBrandRetriever should have been associated with the minecraft jar file, and should have returned us a valid, intact minecraft jar location. This did not work. Either you have modified the minecraft jar file (if so run the forge installer again), or you are using a base editing jar that is changing this class (and likely others too). If you REALLY want to run minecraft in this configuration, add the flag -Dfml.ignoreInvalidMinecraftCertificates=true to the 'JVM settings' in your launcher profile.
```

Judging by the logs and code, you are encountering an issue when launching Minecraft Forge. The main error:  

```
[SEVERE] [ForgeModLoader] The minecraft jar file appears to be corrupt! There has been CRITICAL TAMPERING WITH MINECRAFT
```  

This error message often appears when Forge detects changes in the Minecraft files that it considers suspicious. The parameter `-Dfml.ignoreInvalidMinecraftCertificates=true` tells Forge to ignore certificate verification, which should resolve the issue.

GamerVII-NET replied:
Starting from Minecraft version 1.6.2 - 1.7.2, the game won't launch without this flag.

AlphaBs replied:
okay. I will document it.

AlphaBs commented on SampleForgeInstaller/Program.cs:
```diff
@@ -41,9 +53,13 @@
 };
 
-var process = await launcher.CreateProcessAsync(version_name, launchOption);
+var process = await launcher.BuildProcessAsync(version_name, launchOption);
```
BuildProcessAsync is not the same as CreateProcessAsync. use InstallAndBuildProcessAsync instead. [src](https://github.com/CmlLib/CmlLib.Core/blob/7f22116bf20c6b97247cf673b79ef73b68362e8c/src/MinecraftLauncher.cs#L221)

GamerVII-NET replied:
I have tested and verified that everything works correctly using CreateProcessAsync and BuildProcessAsync for Minecraft versions 1.5.2, 1.6.1, 1.6.2, 1.6.4, and 1.7.2 with their respective Forge versions. Everything functions as expected.

AlphaBs replied:
it won't work if some files are deleted. delete <gamedir>/versions/1.5.2/1.5.2.jar file and launch again. it won't work. `InstallAsync` method ensures that all files exists.

GamerVII-NET replied:
> it won't work if some files are deleted. delete /versions/1.5.2/1.5.2.jar file and launch again. it won't work. `InstallAsync` method ensures that all files exists.

Yes, I understand that. I sent `BuildProcessAsync` by mistake, as it doesn’t handle the installation. It was necessary for my testing purposes. I’ve tested launching with `CreateProcessAsync` — and it works as expected.

Gamer-VII committed:
Remove unused extractUniversal method from FOldest.cs

The extractUniversal method was commented out and no longer in use. Cleaning up this dead code improves maintainability and reduces clutter in the codebase.

Gamer-VII committed:
Rename BuildProcessAsync to CreateProcessAsync for correctly

AlphaBs commented:
your code looks fine. after some test I will merge it soon. great!

AlphaBs commented:
do you know this error? when I try to install 1.7.2 forge after deleting .minecraft directory it occurs:

```
[13:06:43] [main/INFO] [LaunchWrapper]: Loading tweak class name cpw.mods.fml.common.launcher.FMLTweaker
[13:06:43] [main/INFO] [LaunchWrapper]: Using primary tweak class name cpw.mods.fml.common.launcher.FMLTweaker
[13:06:43] [main/INFO] [LaunchWrapper]: Calling tweak class cpw.mods.fml.common.launcher.FMLTweaker
[13:06:43] [main/INFO] [FML]: Forge Mod Loader version 7.2.217.1161 for Minecraft 1.7.2 loading
[13:06:43] [main/INFO] [FML]: Java is Java HotSpot(TM) 64-Bit Server VM, version 1.8.0_51, running on Windows 10:amd64:10.0, installed at C:\Users\ksi12\AppData\Roaming\.minecraft\runtime\windows-x64\jre-legacy
[13:06:43] [main/INFO] [LaunchWrapper]: Loading tweak class name cpw.mods.fml.common.launcher.FMLInjectionAndSortingTweaker
[13:06:43] [main/INFO] [LaunchWrapper]: Loading tweak class name cpw.mods.fml.common.launcher.FMLDeobfTweaker
[13:06:43] [main/INFO] [LaunchWrapper]: Calling tweak class cpw.mods.fml.common.launcher.FMLInjectionAndSortingTweaker
[13:06:43] [main/ERROR] [LaunchWrapper]: Unable to launch
java.util.ConcurrentModificationException
        at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901) ~[?:1.8.0_51]
        at java.util.ArrayList$Itr.remove(ArrayList.java:865) ~[?:1.8.0_51]
        at net.minecraft.launchwrapper.Launch.launch(Launch.java:117) [launchwrapper-1.9.jar:?]
        at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.9.jar:?]
```

source code:
```csharp
using CmlLib.Core.Installer.Forge;
using CmlLib.Core;
using CmlLib.Core.Auth;
using CmlLib.Core.Installers;
using CmlLib.Core.ProcessBuilder;

var path = new MinecraftPath(); // use default directory
var launcher = new MinecraftLauncher(path);

// show launch progress to console
var fileProgress = new SyncProgress<InstallerProgressChangedEventArgs>(e =>
    Console.WriteLine($"[{e.EventType}][{e.ProgressedTasks}/{e.TotalTasks}] {e.Name}"));
var byteProgress = new SyncProgress<ByteProgress>(e =>
    Console.WriteLine(e.ToRatio() * 100 + "%"));
var installerOutput = new SyncProgress<string>(Console.WriteLine);

//Initialize variables with the Minecraft version and the Forge version

var mcVersion = "1.7.2";
var forgeVersion = "10.12.2.1161";

//Initialize MForge
var forge = new ForgeInstaller(launcher);

var version_name = await forge.Install(mcVersion, forgeVersion, new ForgeInstallOptions
{
    FileProgress = fileProgress,
    ByteProgress = byteProgress,
    InstallerOutput = installerOutput,
});
//var version_name = await forge.Install(mcVersion); // install the recommended forge version for mcVersion
//OR var version_name = forge.Install(mcVersion, forgeVersion).GetAwaiter().GetResult();

//Start Minecraft
var launchOption = new MLaunchOption
{
    MaximumRamMb = 1024,
    Session = MSession.CreateOfflineSession("TaiogStudio"),
    ExtraJvmArguments = new []
    {
        new MArgument("-Dfml.ignoreInvalidMinecraftCertificates=true"),
    }
};

var process = await launcher.CreateProcessAsync(version_name, launchOption);

// print game logs
var processUtil = new ProcessWrapper(process);
processUtil.OutputReceived += (s, e) => Console.WriteLine(e);
processUtil.StartWithEvents();
await processUtil.WaitForExitTaskAsync();
```
